haystack/agents/base.py

```diff
@@ -0,0 +1,298 @@
+                        `{"Retriever": {"top_k": 10}, "Reader": {"top_k": 3}}`.
+                        Parameters can only be passed to tools that are pipelines but not nodes.
+        """
+        if len(self.tools) == 0:
+            raise AgentError(
+                "Agents without tools cannot be run. Add at least one tool using `add_tool()` or set the parameter `tools` when initializing the Agent."
+            )
+        if max_iterations is None:
+            max_iterations = self.max_iterations
+
+        transcript = ""
+        # Generate string representations of the tools to be passed to the PromptTemplate of the Agent
+        tool_names = ", ".join(self.tools.keys())
+        tool_names_with_descriptions = "\n".join([f"{tool.name}: {tool.description}" for tool in self.tools.values()])
+
+        # Generate a thought with a plan what to do, choose a tool, generate input for it, and run it.
+        # Repeat this until the final answer is found or the maximum number of iterations is reached.
+        for _ in range(max_iterations):
+            pred = self.prompt_node.prompt(
+                prompt_template=self.prompt_template,
+                query=query,
+                tool_names=tool_names,
+                tool_names_with_descriptions=tool_names_with_descriptions,
+                generated_text=transcript,
+            )[0]
+
+            # Try to extract the final answer or a tool name and tool input from the generated text
+            final_answer = self._extract_final_answer(pred=pred)
+            if final_answer is not None:
+                return self._format_answer(
+                    query=query,
+                    answer=final_answer,
+                    tool_names=tool_names,
+                    tool_names_with_descriptions=tool_names_with_descriptions,
+                    generated_text=transcript + pred,
+                )
+            tool_name, tool_input = self._extract_tool_name_and_tool_input(pred=pred)
+            if tool_name is None or tool_input is None:
+                raise AgentError("Wrong output format.")
+
+            result = self._run_tool(tool_name=tool_name, tool_input=tool_input, params=params)
+            observation = self._extract_observation(result)
+            transcript += f"{pred}\nObservation: {observation}\nThought: Now that I know that {observation} is the answer to {tool_name} {tool_input}, I "
+
+        logger.warning(
+            "Maximum number of agent iterations (%s) reached for query (%s). Increase max_iterations "
+            "or no answer can be provided for this query.",
+            max_iterations,
+            query,
+        )
+        return self._format_answer(
+            query=query,
+            answer="",
+            tool_names=tool_names,
+            tool_names_with_descriptions=tool_names_with_descriptions,
+            generated_text=transcript,
+        )
+
```

Comment:
We could invert the if statement on line 157. So if we don't have a final answer - we take a step. And else, if we do, we format that answer and return it. Not sure if possible, but if it is - could save us 6-7 lines of code in calling _format_answer?

Reply:
There is a third case, which needs to be handled if `max_iterations` is reached. So if I use a `break` condition it's enough to call `_format_answer` only once in the code, otherwise twice I think. Happy to discuss when we meet.

Reply:
@vblagoje I would leave the loop as it is now. The call to `_format_answer` is simpler but there are still two calls of that function. Could you please have another look to see whether you could approve the current version?